Refactor fast gradcheck

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56964 Clean up naming and comments
* **#55871 Refactor fast gradcheck**

Differential Revision: [D28096549](https://our.internmc.facebook.com/intern/diff/D28096549)

soulitzer committed:
Refactor fast gradcheck

[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -337,6 +337,17 @@ def get_jvp_wrt_specific_input(fn, input_idx, inputs, outputs, u, eps) -> List[t
+def get_fast_numerical_jacobians(fn, inputs, inp_indices, outputs, all_u, all_v, eps):
+    reduced_jacobians: List[List[torch.Tensor]] = []
+    for i, (inp_idx, u) in enumerate(zip(inp_indices, all_u)):
+        jacobian_cols = get_jvp_wrt_specific_input(fn, inp_idx, inputs, outputs, u, eps)
```
These are not really columns? Maybe `all_Ju` is a better name?

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -419,6 +430,25 @@ def vjp_fn(grad_output):
+def get_fast_analytical_jacobians(inputs, outputs, nondet_tol, check_grad_dtypes, all_v, all_u_dense):
+    reduced_jacobians: List[List[torch.Tensor]] = []
+    for output, v in zip(outputs, all_v):
+        jacobian_cols = check_analytical_jacobian_attributes(inputs, output, nondet_tol, check_grad_dtypes,
```
Same about these not really being columns

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -419,6 +430,25 @@ def vjp_fn(grad_output):
+                                                             fast_mode=True, v=v)
+        jacobian_scalars: List[torch.Tensor] = []
+        for vJ, u in zip(jacobian_cols, all_u_dense):
+            vJ = vJ.T.squeeze(0)
```
This squeeze looks very arbitrary. Why do you need it?

soulitzer replied:
vJ is a 2-d tensor because this allows us to reuse the error checking logic in the slow case as-is

albanD replied:
Ok, Maybe an extra comment would help.

albanD reviewed (commented):
LGTM
Some variable naming could be improved I think, but good otherwise!

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

albanD reviewed (approved):
LGTM

soulitzer committed:
Update on "Refactor fast gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"


Differential Revision: [D28096549](https://our.internmc.facebook.com/intern/diff/D28096549)

[ghstack-poisoned]

soulitzer committed:
Update on "Refactor fast gradcheck"


Differential Revision: [D28096549](https://our.internmc.facebook.com/intern/diff/D28096549)

[ghstack-poisoned]
